rust/arrow/src/array.rs

```diff
@@ -975,7 +970,7 @@ impl Array for StructArray {
 
     /// Returns the length (i.e., number of elements) of this array
     fn len(&self) -> usize {
-        self.boxed_fields[0].len()
+        self.data().len()
```

Comment:
@sunchao this fails the CI which indicates deeper issues, because `boxed_fields` serves only as a cache (at least in the cpp impl)

Reply:
I was struggling to see what was causing struct-array tests to fail, thanks @kszucs

Reply:
Hi @kszucs @sunchao, can a `StructArray` ever have a null value, or would it always be 0?

I found the issue here, when we create `StructArray` in `impl From<Vec<(Field, ArrayRef)>> for StructArray`, we check that all arrays have the same length, but never assign that length to the array buffer.

Setting a value for the length fixes the failure, but I'm not sure of what to do with null value. I'll check cpp codebase in the mean time.

```rust
[arrow\examples\dynamic_types.rs:62] &nested.data() = ArrayData {
    data_type: Struct(
        [
            Field {
                name: "a",
                data_type: Utf8,
                nullable: false
            },
            Field {
                name: "b",
                data_type: Float64,
                nullable: false
            },
            Field {
                name: "c",
                data_type: Float64,
                nullable: false
            }
        ]
    ),
    len: 5,
    null_count: 0,
    offset: 0,
    buffers: [],
    child_data: [
        ArrayData {
            data_type: Utf8,
            len: 5,
            null_count: 0,
            offset: 0,
            buffers: [
                Buffer {
                    data: BufferData {
                        ptr: 0x0000022cf1290e80,
                        len: 24
                    },
                    offset: 0
                },
                Buffer {
                    data: BufferData {
                        ptr: 0x0000022cf1290f00,
                        len: 5
                    },
                    offset: 0
                }
            ],
            child_data: [],
            null_bitmap: None
        },
        ArrayData {
            data_type: Float64,
            len: 5,
            null_count: 0,
            offset: 0,
            buffers: [
                Buffer {
                    data: BufferData {
                        ptr: 0x0000022cf1291840,
                        len: 40
                    },
                    offset: 0
                }
            ],
            child_data: [],
            null_bitmap: None
        },
        ArrayData {
            data_type: Float64,
            len: 5,
            null_count: 0,
            offset: 0,
            buffers: [
                Buffer {
                    data: BufferData {
                        ptr: 0x0000022cf1291940,
                        len: 40
                    },
                    offset: 0
                }
            ],
            child_data: [],
            null_bitmap: None
        }
    ],
    null_bitmap: None
}
```

Reply:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/array.cc#L373 indicates that this is possible, but I'm not getting joy from the docs (pyarrow and cpp)

Reply:
@nevi-me yes StructArray can have null values - see https://arrow.apache.org/docs/format/Layout.html#struct-type.